### PR TITLE
Core: Add tests for catalogs supporting empty namespaces

### PR DESCRIPTION
**Pull request by nastra**
From all the catalogs that extend `CatalogTests`, only Hive + REST don't support creating a table/view and also listing tables/views in an empty namespace.

JDBC + InMemory do support that and therefore override the `supportsEmptyNamespace()` feature in their tests.

Additionally, this PR also fixes listing tables/views for InMemory as previously this would list all available tables/views, instead of just the ones in that namespace.

I've decided to not adjust the Nessie tests, since the behavior deviates from the current behavior that can be found for JDBC + InMemory. While Nessie doesn't support creation/removal of empty namespaces, it does support creating a table/view in an empty namespace. Listing tables/views in an empty namespace returns all available tables/views.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java [hunk 1]
```diff
@@ -305,9 +305,12 @@ public List<TableIdentifier> listTables(Namespace namespace) {
     }
 
     return fetch(
-        row ->
-            JdbcUtil.stringToTableIdentifier(
-                row.getString(JdbcUtil.TABLE_NAMESPACE), row.getString(JdbcUtil.TABLE_NAME)),
+        row -> {
+          String ns = row.getString(JdbcUtil.TABLE_NAMESPACE);
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java:309]: previously this would return the identifier as `.tblName` instead of `tblName` when using an empty namespace

File: core/src/test/java/org/apache/iceberg/jdbc/TestJdbcCatalog.java [hunk 2]
```diff
@@ -102,12 +102,12 @@ protected JdbcCatalog catalog() {
   }
 
   @Override
-  protected boolean supportsNamespaceProperties() {
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/jdbc/TestJdbcCatalog.java:105]: Intellij flags this because `supportsNamespaceProperties()` is `true` by default in `CatalogTests`, so no need to override this method here

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java [hunk 3]
```diff
@@ -305,16 +305,21 @@ public List<TableIdentifier> listTables(Namespace namespace) {
     }
 
     return fetch(
-        row ->
-            JdbcUtil.stringToTableIdentifier(
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalog.java:309]: previously this would return the identifier as `.tblName` instead of `tblName` when using an empty namespace

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcUtil.java [hunk 4]
```diff
@@ -468,6 +468,10 @@ private JdbcUtil() {}
 
   static Namespace stringToNamespace(String namespace) {
     Preconditions.checkArgument(namespace != null, "Invalid namespace %s", namespace);
+    if (namespace.isEmpty()) {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/jdbc/JdbcUtil.java:471]: previously this method would return the identifier as `.tblName` instead of `tblName` when using an empty namespace, because L475 passes an array with a single empty string to `Namespace.of()`. 

I guess the question here is: is `Namespace.of("")` the same as `Namespace.empty()` and should we maybe fix `Namespace.of()`?
It seems there are currently no existing tests that would indicate what the correct behavior should be in this case

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java [hunk 5]
```diff
@@ -987,6 +991,38 @@ public void testListTables() {
     assertEmpty("Should not contain ns_2.table_1 after drop", catalog, ns2);
   }
 
+  @Test
+  public void emptyNamespace() {
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:995]: added this test because JDBC was previously returning namespace `""` (empty string) when listing all namespaces

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieCatalog.java [hunk 6]
```diff
@@ -169,4 +171,25 @@ public void testWarehouseLocationWithTrailingSlash() {
...
+
+    TableIdentifier table1 = TableIdentifier.of(Namespace.empty(), "table_1");
+    TableIdentifier table2 = TableIdentifier.of(ns, "table_2");
+
+    catalog().buildTable(table1, SCHEMA).create();
+    catalog().buildTable(table2, SCHEMA).create();
+
+    // TODO this lists all available tables and deviates from the behavior in other catalogs and
+    // should probably be fixed
+    Assertions.assertThat(catalog().listTables(Namespace.empty()))
```
Comment (ajantha-bhat) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieCatalog.java:192]: @snazy, @adutra, @dimas-b: Nessie doesn't support empty namespace creation. So, should we allow listing tables from empty namespace?

Comment (nastra) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieCatalog.java:192]: But nessie supports creating a table/view withing the empty namespace, which is a valid use case. If creating a table/view in an empty namespace is supported, then listing should be supported as well

File: core/src/main/java/org/apache/iceberg/inmemory/InMemoryCatalog.java [hunk 7]
```diff
@@ -395,7 +395,9 @@ public void doCommit(TableMetadata base, TableMetadata metadata) {
       String oldLocation = base == null ? null : base.metadataFileLocation();
 
       synchronized (InMemoryCatalog.this) {
-        if (null == base && !namespaceExists(tableIdentifier.namespace())) {
+        if (null == base
+            && !namespaceExists(tableIdentifier.namespace())
+            && !tableIdentifier.namespace().isEmpty()) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/inmemory/InMemoryCatalog.java:400]: Why do we need to check for empty here? Shouldn't this be consistent with `namespaceExists` so that method covers it?

Comment (nastra) [reply core/src/main/java/org/apache/iceberg/inmemory/InMemoryCatalog.java:400]: that's similar to how it's done in https://github.com/apache/iceberg/blob/2446cee5cf0ad93a2be9a68f0b2f7f6fa6edb865/core/src/main/java/org/apache/iceberg/inmemory/InMemoryCatalog.java#L141, where the empty namespace is allowed for listing

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java [hunk 8]
```diff
@@ -168,6 +168,14 @@ protected boolean supportsNamesWithDot() {
     return true;
   }
 
+  protected boolean supportsCreatingEmptyNamespace() {
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:171]: currently, the only reason that two flags exist is because Nessie's behavior is different from JDBC + InMemory.

File: core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java [hunk 9]
```diff
@@ -987,6 +995,61 @@ public void testListTables() {
     assertEmpty("Should not contain ns_2.table_1 after drop", catalog, ns2);
   }
 
+  @Test
+  public void listNamespacesWithEmptyNamespace() {
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:999]: this behavior applies for all catalogs

File: core/src/main/java/org/apache/iceberg/inmemory/InMemoryCatalog.java (hunk 7 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/inmemory/InMemoryCatalog.java]: I've reverted this change as it should actually require that the empty namespace has been created first

File: core/src/main/java/org/apache/iceberg/inmemory/InMemoryCatalog.java [hunk 10]
```diff
@@ -288,6 +288,7 @@ public List<Namespace> listNamespaces(Namespace namespace) throws NoSuchNamespac
 
     List<Namespace> filteredNamespaces =
         namespaces.keySet().stream()
+            .filter(n -> !n.isEmpty())
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/inmemory/InMemoryCatalog.java:291]: this would previously fail in L303 when calling `listNamespaces(Namespace.empty())`

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieCatalog.java (hunk 6 above)
Comment (dimas-b) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieCatalog.java]: An empty namespace is not an "entity", right? Why would we want to "create" it?

That said, creating tables in the "empty" / "root" namespace should work with Nessie, IIRC.

File: core/src/main/java/org/apache/iceberg/inmemory/InMemoryCatalog.java (hunk 10 above)
Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/inmemory/InMemoryCatalog.java:291]: So does this one support empty namespaces or not?

File: core/src/main/java/org/apache/iceberg/jdbc/JdbcUtil.java (hunk 4 above)
Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/jdbc/JdbcUtil.java:471]: I think they should be treated the same, I also hope it doesn't come up very often

File: core/src/main/java/org/apache/iceberg/inmemory/InMemoryCatalog.java (hunk 10 above)
Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/inmemory/InMemoryCatalog.java:291]: Just wondering what the behavior should be here, since it is a valid namespace so shouldn't we be able to list it?

Comment (nastra) [reply core/src/main/java/org/apache/iceberg/inmemory/InMemoryCatalog.java:291]: yes, `InMemoryCatalog` supports empty namespaces and this comes back to the TODO I added in https://github.com/apache/iceberg/pull/9890/files#diff-de393039ffcf049ebb22892ac1b9aad88d53274c1e86431b45e776e4568246f8R1037 to determine what the listing behavior with empty namespaces should be

File: core/src/test/java/org/apache/iceberg/jdbc/TestJdbcUtil.java [hunk 11]
```diff
@@ -45,4 +47,10 @@ public void testFilterAndRemovePrefix() {
 
     assertThat(expected).isEqualTo(actual);
   }
+
+  @Test
+  public void emptyNamespaceInIdentifier() {
```
Comment (RussellSpitzer) [inline core/src/test/java/org/apache/iceberg/jdbc/TestJdbcUtil.java:52]: 👍

File: spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestViews.java [hunk 12]
```diff
@@ -1429,40 +1429,28 @@ public void showViews() throws NoSuchTableException {
...
-    assertThat(sql("SHOW VIEWS"))
-        .contains(
-            row(NAMESPACE.toString(), "prefixV2", false),
-            row(NAMESPACE.toString(), "prefixV3", false),
-            row(NAMESPACE.toString(), "v1", false),
-            tempView);
+    Object[] v1 = row(NAMESPACE.toString(), "v1", false);
+    Object[] v2 = row(NAMESPACE.toString(), "prefixV2", false);
+    Object[] v3 = row(NAMESPACE.toString(), "prefixV3", false);
+    assertThat(sql("SHOW VIEWS")).contains(v2, v3, v1, tempView);
```
Comment (RussellSpitzer) [inline spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestViews.java:1435]: the order here

**Review by RussellSpitzer — APPROVED** [review]
LGTM, only small issue with test rows being defined, 1, 2, 3 then used 2, 3, 1. I keep thinking there is a secret there but I know there isn't.

**Review by Fokko — APPROVED** [review]
Looks good, thanks @nastra. Thanks @RussellSpitzer for the review, let's get this in and follow up on the TODO in a separate PR
